### ISSUE TITLE
crypto/tls: add fields to ClientHelloInfo

### ISSUE THREAD

**Issue by FiloSottile** [golang/go#17430]
ClientHelloInfo is how callbacks make decisions on how to treat an incoming connection, which is becoming very powerful with https://golang.org/cl/30790/

As mentioned in https://github.com/golang/go/issues/16066#issuecomment-250608294 it would be useful to have some more information in there. 

The Cloudflare fork exposes Version, SignatureSchemes and LocalAddr. RemoteAddr seems fitting for completeness. ALPNProtocols as asked in #16066.

Looking at clientHelloMsg, only session resumption information, ocspStapling and scts are left, for which I can't think of a use case.

**Comment by titanous** [comment]
> Looking at clientHelloMsg, only session resumption information, ocspStapling and scts are left, for which I can't think of a use case.

It would be useful to have access to all of the information provided in the ClientHello to collect statistics on handshakes.

**Comment by FiloSottile** [comment]
@titanous Something we did again in the Cloudflare fork is expose the serialized CH in the ConnectionState, but I don't think anything that low-level fits in the standard library. Or did you have anything else in mind?

**Comment by titanous** [comment]
> Something we did again in the Cloudflare fork is expose the serialized CH in the ConnectionState, but I don't think anything that low-level fits in the standard library. Or did you have anything else in mind?

It's not unprecedented. [`x509.Certificate`](https://golang.org/pkg/crypto/x509/#Certificate) has a bunch of `Raw*` fields. Another option would be exposing all parameters/extensions that the package is aware of and then using a single field for all unknown extensions.

**Comment by FiloSottile** [comment]
True, but then I would go for having a high-level CHI, and a RawClientHello, rather than polluting CHI with sort-of-raw fields. Opinions on whether it should go on CHI or on ConnState? We wanted it in ConnState so we could observe it from a http.Handler. Maybe ConnectionState should expose a CHI?

Also, in 1.3 a CH might be a significant chunk of memory to keep around.

**Comment by FiloSottile** [comment]
Here is my patch on top of https://golang.org/cl/30790/, which I can't submit yet because the Go Gerrit does not seem to support CL with dependencies on CLs by other authors.

SupportedVersions is terrible, but needed to be forward-compatible with the TLS 1.3 extension-based version negotiation. The SignatureSchemes term is stolen from 1.3 and is uint16 to be forward-compatible.

/cc @agl 

```diff
commit 3166b277d4be70a0b0f9864a0751a757cea7761b
Author: Filippo Valsorda <hi@filippo.io>
Date:   Thu Oct 13 15:46:06 2016 +0100

    crypto/tls: expand ClientHelloInfo
    
    Fixes #17430
    
    Change-Id: Ia1c25363d64e3091455ce00644438715aff30a0d

diff --git a/src/crypto/tls/common.go b/src/crypto/tls/common.go
index f20fb1b..9507c92 100644
--- a/src/crypto/tls/common.go
+++ b/src/crypto/tls/common.go
@@ -14,6 +14,7 @@ import (
 	"fmt"
 	"io"
 	"math/big"
+	"net"
 	"strings"
 	"sync"
 	"time"
@@ -217,6 +218,11 @@ type ClientSessionCache interface {
 // ClientHelloInfo contains information from a ClientHello message in order to
 // guide certificate selection in the GetCertificate callback.
 type ClientHelloInfo struct {
+	// SupportedVersions lists the TLS versions supported by the client.
+	// In current TLS this is extrapolated from the max version advertised by
+	// the client, so values that are not the highest might be rejected if used.
+	SupportedVersions []uint16
+
 	// CipherSuites lists the CipherSuites supported by the client (e.g.
 	// TLS_RSA_WITH_RC4_128_SHA).
 	CipherSuites []uint16
@@ -238,6 +244,26 @@ type ClientHelloInfo struct {
 	// is being used (see
 	// http://tools.ietf.org/html/rfc4492#section-5.1.2).
 	SupportedPoints []uint8
+
+	// SignatureSchemes lists the signature and hash schemes that the client
+	// is willing to verify. SignatureSchemes is set only if the Signature
+	// Algorithms Extension is being used (see
+	// https://tools.ietf.org/html/rfc5246#section-7.4.1.4.1).
+	SignatureSchemes []uint16
+
+	// SupportedProtos lists the application protocols supported by the client.
+	// SupportedPoints is set only if the Application-Layer Protocol
+	// Negotiation Extension is being used (see
+	// https://tools.ietf.org/html/rfc7301#section-3.1).
+	//
+	// Servers can select a protocol by setting Client.NextProtos in a
+	// GetConfigForClient return value.
+	SupportedProtos []string
+
+	// LocalAddr is the address on which the connection was accepted.
+	LocalAddr net.Addr
+	// RemoteAddr is the client address.
+	RemoteAddr net.Addr
 }
 
 // RenegotiationSupport enumerates the different levels of support for TLS
diff --git a/src/crypto/tls/handshake_server.go b/src/crypto/tls/handshake_server.go
index 52d7de9..571f5c0 100644
--- a/src/crypto/tls/handshake_server.go
+++ b/src/crypto/tls/handshake_server.go
@@ -124,11 +124,28 @@ func (hs *serverHandshakeState) readClientHello() (isResume bool, err error) {
 		return false, unexpectedMessageError(hs.clientHello, msg)
 	}
 
+	supportedVersions := []uint16{}
+	for _, v := range []uint16{VersionSSL30, VersionTLS10, VersionTLS11, VersionTLS12} {
+		if v <= hs.clientHello.vers {
+			supportedVersions = append(supportedVersions, v)
+		} else {
+			break
+		}
+	}
+	var signatureSchemes []uint16
+	for _, sah := range hs.clientHello.signatureAndHashes {
+		signatureSchemes = append(signatureSchemes, uint16(sah.hash)<<8+uint16(sah.signature))
+	}
 	clientHelloInfo := &ClientHelloInfo{
-		CipherSuites:    hs.clientHello.cipherSuites,
-		ServerName:      hs.clientHello.serverName,
-		SupportedCurves: hs.clientHello.supportedCurves,
-		SupportedPoints: hs.clientHello.supportedPoints,
+		SupportedVersions: supportedVersions,
+		CipherSuites:      hs.clientHello.cipherSuites,
+		ServerName:        hs.clientHello.serverName,
+		SupportedCurves:   hs.clientHello.supportedCurves,
+		SupportedPoints:   hs.clientHello.supportedPoints,
+		SignatureSchemes:  signatureSchemes,
+		SupportedProtos:   hs.clientHello.alpnProtocols,
+		LocalAddr:         hs.c.LocalAddr(),
+		RemoteAddr:        hs.c.RemoteAddr(),
 	}
 
 	if config.GetConfigForClient != nil {

```

**Comment by bradfitz** [comment]
@agl, decision?
